src/tiled/objecttypesmodel.cpp

```diff
@@ -91,8 +97,17 @@ bool ObjectTypesModel::setData(const QModelIndex &index,
                                int role)
 {
     if (role == Qt::EditRole && index.column() == 0) {
-        mObjectTypes[index.row()].name = value.toString().trimmed();
+        const int oldRow = index.row();
+        mObjectTypes[oldRow].name = value.toString().trimmed();
         emit dataChanged(index, index);
+
+        auto nextObjectType = std::lower_bound(mObjectTypes.begin(), mObjectTypes.end(), mObjectTypes[oldRow], objectTypeLessThan);
+        const int newRow = nextObjectType - mObjectTypes.begin();
+        if (newRow != oldRow) {
+            beginMoveRows(QModelIndex(), oldRow, oldRow, QModelIndex(), newRow);
+            mObjectTypes.move(oldRow, newRow > oldRow ? newRow - 1 : newRow);
```

Comment:
It makes sense if you consider that `move(from,to)` is just a more efficient way of doing `insert(to,takeAt(from))`. After the operation, you expect the element's index to be `to`.